resources/virt.go

```diff
@@ -0,0 +1,521 @@
+		fmt.Println("Connection to libvirt faild")
+		return err
+	}
+	obj.conn = conn
+
+	obj.BaseRes.Init() // call base init, b/c we're overriding
+
+	return nil
+}
+
+func (obj *VirtRes) Validate() bool {
+	return true
+}
+
+func (obj *VirtRes) GetUUIDs() []ResUUID {
+	x := &VirtUUID{
+		BaseUUID: BaseUUID{name: obj.GetName(), kind: obj.Kind()},
+	}
+	return []ResUUID{x}
+}
+
+func (obj *VirtRes) Watch(processChan chan event.Event) error {
+	if obj.IsWatching() {
+		return nil
+	}
+	obj.SetWatching(true)
+	defer obj.SetWatching(false)
+	cuuid := obj.converger.Register()
+	defer cuuid.Unregister()
+
+	ch := make(chan int, 10)
+	callback := libvirt.DomainEventCallback(
+		func(c *libvirt.VirConnection, d *libvirt.VirDomain, eventDetails interface{}, f func()) int {
+			if lifecycleEvent, ok := eventDetails.(libvirt.DomainLifecycleEvent); ok {
+				domName, _ := d.GetName()
+				if domName == obj.BaseRes.Name {
+					ch <- lifecycleEvent.Event
+				}
+			} else {
+				fmt.Println("event details isn't DomainLifecycleEvent")
+			}
+			return 0
+		},
+	)
+	callbackId := obj.conn.DomainEventRegister(
+		libvirt.VirDomain{},
+		libvirt.VIR_DOMAIN_EVENT_ID_LIFECYCLE,
+		&callback,
+		nil,
+	)
+	defer obj.conn.DomainEventDeregister(callbackId)
+
+	var send = false
+	var exit = false
+	var dirty = false
+
+	for {
+		select {
+		case event := <-ch:
+			switch event {
+			case libvirt.VIR_DOMAIN_EVENT_DEFINED:
+				if obj.Transient {
+					dirty= true
+					send = true
+				}
+			case libvirt.VIR_DOMAIN_EVENT_UNDEFINED:
+				if !obj.Transient {
+					dirty= true
+					send = true
+				}
+			case libvirt.VIR_DOMAIN_EVENT_STARTED:
+				fallthrough
+			case libvirt.VIR_DOMAIN_EVENT_RESUMED:
+				if obj.State != "running" {
+					dirty= true
+					send = true
+				}
+			case libvirt.VIR_DOMAIN_EVENT_SUSPENDED:
+				if obj.State != "paused" {
+					dirty= true
+					send = true
+				}
+			case libvirt.VIR_DOMAIN_EVENT_STOPPED:
+				fallthrough
+			case libvirt.VIR_DOMAIN_EVENT_SHUTDOWN:
+				if obj.State != "shutoff" {
+					dirty= true
+					send = true
+				}
+			case libvirt.VIR_DOMAIN_EVENT_PMSUSPENDED:
+				fallthrough
+			case libvirt.VIR_DOMAIN_EVENT_CRASHED:
+				dirty= true
+				send = true
+			}
+
+		case event := <-obj.events:
+			cuuid.SetConverged(false)
+			if exit, send = obj.ReadEvent(&event); exit {
+				return nil// exit
+			}
+
+		case _ = <-cuuid.ConvergedTimer():
+			cuuid.SetConverged(true) // converged!
+			continue
+		}
+
+		if send {
+			send = false
+			if dirty {
+				dirty = false
+				obj.isStateOK = false // something made state dirty
+			}
+			resp := event.NewResp()
+			processChan <- event.Event{event.EventNil, resp, "", true} // trigger process
+			resp.ACKWait()                                 // wait for the ACK()
+		}
+	}
+}
+
+func (obj *VirtRes) CheckApply(apply bool) (bool, error) {
+	log.Printf("%v[%v]: CheckApply(%t)", obj.Kind(), obj.GetName(), apply)
+
+	if obj.isStateOK { // cache the state
```

Comment:
This isn't really being used ATM, but probably should be. Basically a watch event invalidates the cached event state, but a successful CheckApply should `true` it.